src/main/scala/ir/Expr.scala

```diff
@@ -350,6 +351,9 @@ case class Memory(name: String, addressSize: Int, valueSize: Int) extends Expr w
 sealed trait Variable extends Expr {
   val name: String
   val irType: IRType
+  val ssa_id: mutable.Set[Int] = mutable.Set[Int]()
+  var sharedVariable: Boolean = false
```

Comment:
Making the Variable trait mutable is really not a good way to do this. Exprs are supposed to be immutable. This will be very messy and unpredictable.

Reply:
Perhaps a new SSAVariable case class that contains a Variable and an Int would be a better way to implement SSA?

Reply:
This also needs commenting - what does it mean for a variable to have multiple SSA ids?

Reply:
It's supposed to work like a definitions set, the id is incremented and assigned at every definition, and after a join the variable will contain the ids of the last definition in both paths. Hence the intersection of the set between a variable at different points being nonempty tells you they use the same definition, similar to a phi node.  I think at this point it would be simpler and more reliable to implement a regular usedef analysis with the tip framework with all the state stored externally in a map lattice (cfgposition -> variable -> definition set). We can then optionally use that to transform the IR into an SSA form; maybe wrapping variables in an SSAVariable case class as you suggested; that way its immutable. @yousifpatti